src/client/react.js

```diff
@@ -225,26 +244,25 @@ export function SessionProvider(props) {
     }
 
     __NEXTAUTH._getSession()
-  }, [staleTime])
+  }, [staleTime, broadcastSession])
 
   React.useEffect(() => {
     // Listen for storage events and update session if event fired from
     // another window (but suppress firing another event to avoid a loop)
     // Fetch new session data but tell it to not to fire another event to
     // avoid an infinite loop.
-    // Note: We could pass session data through and do something like
-    // `setData(message.data)` but that can cause problems depending
-    // on how the session object is being used in the client; it is
-    // more robust to have each window/tab fetch it's own copy of the
-    // session object rather than share it across instances.
-    const unsubscribe = broadcast.receive(
-      async () => await __NEXTAUTH._getSession({ event: "storage" })
-    )
-
-    return () => unsubscribe()
-  }, [])
+    return broadcast.receive(async ({ reason }) => {
+      if (
+        (reason !== "signOut" && broadcastSession) ||
+        (reason === "signOut" && broadcastSignOut)
```

Comment:
The logic here seems error prone.
`reason !== 'signOut'` means that it will be true when the reason is `getSession` AND every other reason we might add later.
Should it be instead `reason === 'getSession' && broadcastSession` so we know it's only for the `getSession` event and we don't introduce accidental bugs if we add more reasons later?

That said, when using these constant values I think it would be nice to declare them as constant in order to avoid little mistakes like `signout` vs `signOut`
```js
export const SIGNOUT = 'signOut'
export const GET_SESSION = 'getSession'

// and then always use the constants rather than the string
```